core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -470,11 +469,13 @@ protected final void disconnectFromNodeChannel(final Channel channel, final Exce
     }
 
     protected Channel nodeChannel(DiscoveryNode node, TransportRequestOptions options) throws ConnectTransportException {
-        NodeChannels nodeChannels = connectedNodes.get(node);
-        if (nodeChannels == null) {
-            throw new NodeNotConnectedException(node, "Node not connected");
+        try (Releasable ignored = connectionLock.acquire(node.getId())) {
```

Comment:
This is needed to establish a happens-before edge between `#nodeChannel()` and `#disconnectFromNode()`. Without the lock a connect can sneak in in the middle of a disconnect.